runtime/common/src/migrations/balances.rs

```diff
@@ -0,0 +1,138 @@
+	/// which are suspendable. This includes named reserve and unnamed reserve.
+	pub reserved: Balance,
+	/// The amount that `free` may not drop below when withdrawing for *anything
+	/// except transaction fee payment*.
+	pub misc_frozen: Balance,
+	/// The amount that `free` may not drop below when withdrawing specifically
+	/// for transaction fee payment.
+	pub fee_frozen: Balance,
+}
+
+pub type OldAccountInfo<Index, Balance> = AccountInfo<Index, OldAccountData<Balance>>;
+
+pub type NewAccountData<Balance> = pallet_balances::AccountData<Balance>;
+
+pub type NewAccountInfo<Index, Balance> = AccountInfo<Index, NewAccountData<Balance>>;
+
+pub struct Migration<T: pallet_balances::Config + frame_system::Config>(
+	sp_std::marker::PhantomData<T>,
+);
+
+impl<T> OnRuntimeUpgrade for Migration<T>
+where
+	T: pallet_balances::Config + frame_system::Config,
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		let account_prefix = frame_system::Account::<T>::final_prefix();
+
+		let mut previous_key = account_prefix.to_vec();
+
+		while let Some(next) = sp_io::storage::next_key(&previous_key) {
+			previous_key = next;
+
+			log::info!(
+				"Balances Migration - Processing account key - {}",
+				hex::encode(previous_key.clone())
+			);
+
+			/// The difference between the old and new account data structures
+			/// is the last field of the struct, which is:
+			///
+			/// Old - `free_frozen` - T::Balance
+			/// New - `flags` - u128
+			///
+			/// During this check, we confirm that both the old and the new can
+			/// be successfully decoded given the raw data found in storage, and
+			/// we add specific checks for each version:
+			///
+			/// Old - confirm that `fee_frozen` is zero, as it shouldn't have
+			/// been used so far
+			/// New - confirm that `flags` does not have the
+			/// new logic flag set
+			match unhashed::get::<OldAccountInfo<T::Index, T::Balance>>(&previous_key) {
+				Some(old) => {
+					if !old.data.fee_frozen.is_zero() {
+						log::warn!("Balances Migration - Old account data with non zero frozen fee")
+					}
+				}
+				None => log::error!("Balances Migration - Error decoding old data"),
+			};
+
+			match unhashed::get::<NewAccountInfo<T::Index, T::Balance>>(&previous_key) {
+				Some(new) => {
+					if new.data.flags.is_new_logic() {
+						log::warn!(
+							"Balances Migration - New account data with new logic flag enabled"
+						)
+					}
+				}
+				None => log::error!("Balances Migration - Error decoding new data"),
+			};
+		}
+
+		// CHECKING DECODING OLD DATASTRUCTURE WITH NEW LAYOUT WORKS:
+		// * Fetch storage from chain with NEW data structure
+		// * Check if fetched accounts matches on-chain storage entries
+
+		// TODO:
+		// * Research whether orml-pallets datastructure also changed
+		// * Research whether we need to migrate locks in orml (maybe first check if
+		//   there exist any. ^^
```

Comment:
`orml-tokens` has not updated their `AccountData` yet:
https://github.com/open-web3-stack/open-runtime-module-library/blob/8f59b50ab45d15572d18cee0425966e43567a954/tokens/src/lib.rs#L142-L160

I would assume, they won't do that because to not force consumers into synchronous migrations. In Substrate, the `misc_frozen` field of `AccountData` was replaces with `flags`. However, both fields have the same size which removes the requirement of a sync migration: https://github.com/paritytech/substrate/pull/12951/files#r1140113642

Reply:
Yup, just thought that it might be a good idea to have similar checks like we do for the normal accounts. No actual migrations taking place, just us having some checks before the upgrade itself. WDYT?

Reply:
I think that is a good idea!